autoload/flog.vim

```diff
@@ -1012,6 +1012,190 @@ endfunction
+  else
+    execute "normal! " . v:count1 . "j"
+  endif
+  call flog#to_commit_start()
+endfunction
+
+" See https://vi.stackexchange.com/questions/29062/how-to-check-if-a-string-starts-with-another-string-in-vimscript/29063#29063
+fu! flog#starts_with(longer, shorter) abort
+  return a:longer[0:len(a:shorter)-1] ==# a:shorter
+endfunction
+
+" See https://vi.stackexchange.com/questions/29056/how-to-find-first-item-that-satisfies-predicate/29059#29059
+function! flog#find_predicate(haystack, predicate) abort
+  return get(filter(copy(a:haystack), "a:predicate(v:val)"), 0, v:null)
+endfunction
+
+fu! flog#find_commit(state, commit_hash) abort
+  return flog#find_predicate(a:state.commits, {item -> flog#starts_with(a:commit_hash, item.short_commit_hash)})
+endfunction
+
+fu! flog#jump_to_commit(commit_hash) abort
+  let l:state = flog#get_state()
+  let l:commit = flog#find_commit(l:state, a:commit_hash)
+  if type(l:commit) != v:t_dict
+    let l:state.reflog = v:true
+    call flog#populate_graph_buffer()
+    let l:commit = flog#find_commit(l:state, a:commit_hash)
+  endif
+  let l:index = index(l:state.commits, l:commit)
+  let l:index = min([max([l:index, 0]), len(l:state.commits) - 1])
+  let l:line = index(l:state.line_commits, l:state.commits[l:index]) + 1
+  if l:line >= 0
+    exec l:line
+    call flog#to_commit_start()
+  endif
+endfunction
+
+fu! flog#get_short_commit_hash() abort
+  let l:state = flog#get_state()
+  let l:current_commit = flog#get_commit_at_line()
+  if type(l:current_commit) != v:t_dict
+    return
+  endif
+  return l:current_commit.short_commit_hash
+endfunction
+
+fu! flog#get_full_commit_hash() abort
+  let l:short = flog#get_short_commit_hash()
+  if type(l:short) != v:t_string
+    return v:null
+  endif
+  return fugitive#RevParse(flog#get_short_commit_hash())
+endfunction
+
+" Returns one of the elements of the list.
+" If it has been called before with a commit
+" present in the list, or has previously returned
+" an element of the current list, *that* element will be
+" returned. In addition, if the current commit is
+" not in the list of prior inputs and outputs,
+" that list is cleared. Hence there can only be one
+" matching element.
+" Use this as a replacement for just doing commit_list[0],
+" giving the current commit as an additional input. 
+" In the base case, it just returns
+" the first element. But the most recent inputs and outputs 
+" of this function are given priority before any others,
+" giving the function 'stability' when navigating up/down
+" a directed acyclic graph
+fu! flog#stable_select(commit_list, current_commit) abort
+  if index(g:flog_visited_commits, a:current_commit) == -1
+    let g:flog_visited_commits = [a:current_commit]
+  endif
+  let pick = flog#find_predicate(a:commit_list, {target_commit -> index(g:flog_visited_commits, target_commit) != -1})
+  if pick == v:null
+    let pick = a:commit_list[0]
+  endif
+  call add(g:flog_visited_commits, l:pick)
```

Comment:
History should probably be kept per-repository

Reply:
I'm not too opposed to leaving this for later if this is too much of an ask but let me know what you think because you have a better idea of usability than me right now

Reply:
it would definitely be better per-repository. the behaviour now is that this history is forgotten when you navigate in another repo, because of course the commit hashes won't match with those browsed in the other tab. So that means it will go down/up the wrong path. A minor bug but probably not that hard to fix. Do you already have anything else stored per repo, or should I think of my own solution to get a repo-string?

Reply:
Nothing else stored per-repository. You could use a global object, you could just use something like the Git directory path as the key.

Reply:
I haven't been able to find motivation to work this, because I use this functionality on a daily basis across several repositories, and never notice this being lacking. I am inclined to say it just isn't necessary because the idea is just to make j/k intuitively reversible for when you scroll one too far. Anyway, this could be done in a separate PR.

Reply:
I don't have a problem adding this particular feature on merge myself

Reply:
Fixed; simply using a buffer local variable is enough (in fact, better! two different flog buffers on the same repo will have their own tracking of visited commits)